### DISCUSSION TITLE
How to create Circles with "real" sizes?

### DISCUSSION THREAD

**Discussion by Myzel394** [rnmapbox/maps#1903] [Q&A]
I'm really struggling with this and apparently there literally seems to be no one else wanting this.

How do I make circles that have a "real" size? By "real size" I mean that a given circle has e.g. a radius of 50 meters and then scales accordingly to the current zoom factor, so that it always has a radius of 50 meters (and not 50 pixels).

**[If you have the same issue, see my answer here](https://github.com/rnmapbox/maps/discussions/1903#discussioncomment-2749917)**

**Comment by Gnative** [comment]
Just some ideas that might help you in the right direction:

Add a SymbolLayer layer with styles such as

`iconSize: ['interpolate', ['linear'], ['zoom'], 10, 1, 15, 0.5]`

https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#ramps-scales-curves 

or you could use the zoom event and resize an Animated.ShapeSource based on the zoom ?

**Reply by Gnative** [reply]
I think an issue is that zoom levels have different "meters per pixel " according to latitude and longitude which makes sence I guess with map projections. 
This doc also shows the relation of a meters per pixel
 
https://docs.mapbox.com/help/glossary/zoom-level/

kinda interesting.

**Reply by Gnative** [reply]
Maybe there is something in this which might help too - https://github.com/mapbox/mapbox-gl-js/blob/b32d2ce6f9c59e679464a5ab3719cd1dad7f85a2/src/ui/control/scale_control.js

**Comment by rastapasta** [comment]
@Myzel394 check out turf.js [circle](https://turfjs.org/docs/#circle) method to get what you might want: a polygon around a point with a radius of 50 meters. Just display it via a ```ShapeSource``` & ```FillLayer``` and you should have what you are looking for. Cheers!

**Reply by Myzel394** [reply]
How exactly do I use the `circle` function? I tried this based on a StackOverflow answer (which I can't find anymore, sorry):

```tsx
const circleFeature = circle([1, 1], 5, {
	steps: 50,
	units: "kilometers",
	properties: {foo: "bar"},
})
const feature = turf.lineString(...circleFeature.geometry.coordinates)

/// React component here

return (
    // ...
	<MapboxGL.ShapeSource id="" type="geojson" data={circleFeature} />

	<MapboxGL.CircleLayer
		center={{
			latitude: 1,
			longitude: 1,
		}}
		radius={20}
		strokeWidth={2}
		strokeColor="#3399ff"
		fillColor="#80bfff"
	/>
```

**Reply by rastapasta** [reply]
As given in the linked documentation, ```circle``` returns a  ```Feature <Polygon>``` - give that to turf's ```featureCollection``` to receive a shape which u can pass to [ShapeSource](https://github.com/rnmapbox/maps/blob/main/docs/ShapeSource.md) and display with a nested layer. Please read the corresponding documentations, as all the answers I gave derived from them as well.

![Screenshot 2022-05-13 at 21 19 51](https://user-images.githubusercontent.com/1259904/168375209-d62e9373-9585-486e-8bca-d74827154e12.png)

**Reply by Myzel394** [reply]
Unfortunately, that doesn't work either. Neither when using a `CircleLayer` nor when using a `LineLayer` ([as suggested in this issue response](https://github.com/nitaliano/react-native-mapbox-gl/issues/945#issuecomment-381121954), nothing is shown on the map.

Here's what I tried:

```tsx

const circleFeature = circle([1, 1], 5, {
	steps: 50,
	units: "kilometers",
	properties: {foo: "bar"},
})
const features = featureCollection([circleFeature])

const layerStyles = {
	route: {
		lineColor: "grey",
		lineWidth: 3,
		lineOpacity: 0.84,
		lineDasharray: [2, 2],
	},
}

export default function Map() {
	return (
		<MapboxGL.MapView
			styleURL={MapboxGL.StyleURL.Street}
			style={styles.wrapper}
		>
			<MapboxGL.ShapeSource id="positions" shape={features}>
				<MapboxGL.LineLayer
					id="routeFill"
					style={layerStyles.route}
					belowLayerID="originInnerCircle"
				/>
			</MapboxGL.ShapeSource>
		</MapboxGL.MapView>
	)
}
```

[I also found this StackOverflow answer](https://stackoverflow.com/questions/37599561/drawing-a-circle-with-the-radius-in-miles-meters-with-mapbox-gl-js), but I couldn't find documentation on how to properly transform this from vanilla JavaScript to React Native

**Reply by rastapasta** [reply]
Try to remove the ```belowLayerID``` as it won't get rendered if this layer's id can't be found. Again, please refer to the corresponding [documentation](https://github.com/rnmapbox/maps/tree/main/docs).

**Reply by Myzel394** [reply]
@rastapasta I wouldn't ask (or start this discussion) if the docs were self-explainable.

My original question was: "How to create Circles with "real" sizes?"
So I went to the [CircleLayer docs](https://github.com/rnmapbox/maps/blob/main/docs/CircleLayer.md) - the first thing I read is: "CircleLayer is a style layer that renders one or more filled circles on the map.". Okay.... nice to know but I already guessed that it would do something like that. 
I then checked what props I have to provide and how they are used. Unfortunately, only the props are described. There's no description on how to practically use `CircleLayer`.

How am I supposed to learn anything from that? I can't, I had to lookup at the repo example to at least get an idea on how to use `Source`s and `Layer`s.

And unfortunately removing `belowLayerID` doesn't work either.

**Reply by rastapasta** [reply]
@Myzel394 sorry for the confusion - displaying features via the ```CircleLayers``` and directly displaying polygon features (in this case circle-shaped) are two different concepts. For clarity, following works fine:
![Screenshot 2022-05-13 at 23 58 55](https://user-images.githubusercontent.com/1259904/168394886-a40d0587-c599-4fe8-bfb6-ef5d27b6414c.png)
![Screenshot 2022-05-13 at 23 58 38](https://user-images.githubusercontent.com/1259904/168394887-675baeef-d8c9-44b5-a3e6-8aa824b5b033.png)

**Reply by Myzel394** [reply]
Holy macaroni I finally found that little pesky mistake: My coordinates were in the wrong order!
It's: `[Longitude, Latitude]`!

Given this, I was finally able to create a filled circle the following way:

```tsx
import MapboxGL from "@rnmapbox/maps"
import {StyleSheet} from "react-native"
import {useCallback} from "react"
import circle from "@turf/circle"
import {featureCollection} from "@turf/helpers"

const circleFeature = circle([-122,431297, 37.773972], 3, {
	steps: 60,
	units: "kilometers",
})
const features = featureCollection([circleFeature])

export default function Map() {
	return (
		<MapboxGL.MapView style={styles.wrapper}>
			<MapboxGL.ShapeSource id="positions" shape={features}>
				<MapboxGL.FillLayer
					id="testLayer"
					style={{
						fillColor: "red",
						visibility: "visible",
						fillOpacity: 0.5,
					}}
				/>
			</MapboxGL.ShapeSource>
		</MapboxGL.MapView>
	)
}

const styles = StyleSheet.create({
	wrapper: {
		flex: 1,
	},
})
```

And I think I finally understood how `Source`s and `Layer`s work. `Source`s describe the actual "element" you have, such as a circle, a polygon, etc. And `Layer`s describe what to do with it, add fill to it, add circles around it, etc. It's kinda like HTML and CSS, isn't it?

**Reply by Gnative** [reply]
Cool you found away to do this. Apologies if I put you in the wrong direction earlier.

**Reply by Myzel394** [reply]
No worries. This was actually the first method I tried out (before creating this discussion) - but the method here is better, as it doesn't rely on scaling the circle / element correctly.
You don't need to be sorry for trying to help! :)

**Reply by nesjett** [reply]
I'm a bit late so you already figured it out. For the record, I wrote an article about this and how to make it move with the user location smoothly without it to look glitchy: https://nsabater.com/react-native-mapboxgl-custom-userlocation
